ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,60 +203,429 @@ let order_f (_, vm_rec) =
   let main, shadow =
     Memory_check.vm_compute_start_memory ~__context ~policy snapshot
   in
   Int64.add main shadow
 
-(** Return a VM -> Host plan for the Host.evacuate code. We assume the VMs are all agile. The returned plan may
-    be incomplete if there was not enough memory. *)
-let compute_evacuation_plan ~__context total_hosts remaining_hosts
-    vms_and_snapshots =
-  let hosts =
-    List.map
-      (fun host ->
-        ( host
-        , Memory_check.host_compute_free_memory_with_maximum_compression
-            ~__context ~host None
+(* Used to sort pairs into increasing order of their second component *)
+let compare_snd (_, a) (_, b) = compare a b
+
+let host_free_memory ~__context ~host =
+  Memory_check.host_compute_free_memory_with_maximum_compression ~__context
+    ~host None
+
+let vm_memory ~__context snapshot =
+  let policy =
+    match Helpers.check_domain_type snapshot.API.vM_domain_type with
+    | `hvm | `pv ->
+        Memory_check.Dynamic_min
+    | `pv_in_pvh | `pvh ->
+        Memory_check.Static_max
+  in
+  total_memory_of_vm ~__context policy snapshot
+
+type anti_affinity_pool_state = {
+    host_size_map: int64 Xapi_vm_helpers.HostMap.t
+  ; grp_host_vm_cnt_map: int Xapi_vm_helpers.HostMap.t VMGroupMap.t
+  ; grp_hosts_set: HostsSet.t VMGroupMap.t
+}
+
+let init_host_size_map hosts =
+  List.fold_left
+    (fun m (h, h_size) -> Xapi_vm_helpers.HostMap.add h h_size m)
+    Xapi_vm_helpers.HostMap.empty hosts
+
+let update_host_size_map host vm_size host_size_map =
+  match Xapi_vm_helpers.HostMap.mem host host_size_map with
+  | false ->
+      error "Failed to find free memory of host in host_size_map for host." ;
+      raise (Api_errors.Server_error (Api_errors.internal_error, []))
+  | true ->
+      Xapi_vm_helpers.HostMap.update host
+        (fun host_size ->
+          Int64.sub (Option.value ~default:0L host_size) vm_size |> Option.some
         )
-      )
-      remaining_hosts
+        host_size_map
+
+let anti_affinity_grps ~__context =
+  Db.VM_group.get_all ~__context
+  |> List.filter (fun group ->
+         Db.VM_group.get_placement ~__context ~self:group = `anti_affinity
+     )
+
+let init_anti_affinity_grp_host_vm_cnt_map ~__context =
+  anti_affinity_grps ~__context
+  |> List.fold_left
+       (fun grp_to_host_vm_cnt_map grp ->
+         VMGroupMap.add grp
+           (Xapi_vm_helpers.host_to_vm_count_map ~__context grp)
+           grp_to_host_vm_cnt_map
+       )
+       VMGroupMap.empty
+
+let update_anti_affinity_grp_host_vm_cnt_map group host grp_host_vm_cnt_map =
+  VMGroupMap.update group
+    (fun host_vm_cnt_map ->
+      Some
+        (Xapi_vm_helpers.HostMap.update host
+           (fun vm_cnt -> Option.(value ~default:0 vm_cnt |> succ |> some))
+           (Option.value ~default:Xapi_vm_helpers.HostMap.empty host_vm_cnt_map)
+        )
+    )
+    grp_host_vm_cnt_map
+
+let init_anti_affinity_grp_hosts_set ~__context grp_host_vm_cnt_map hosts =
```

Comment:
I think this performs the following mapping:
```
B: GroupMap: group -> (HostMap : host -> vm_cnt)
->
C: GroupMap: group -> (vm_cnt, free_mem, h)
```
So it could be just as the `B` is already a map, a simple map will do the trick.
```
GroupMap.map (fun hosts ->
  HostMap.to_seq
  |> Seq.map (fun (h, vm_cnt) -> (vm_cnt, (HostMap.find h free_mems), h))
  |> HostSet.of_seq
) GroupMap.empty
```

Reply:
There is no entry for host in the map B when `vm_cnt` of the host is 0 .

Reply:
Could it just be `0` with keys?

Reply:
Currently it is calculated by `Xapi_vm_helpers.host_to_vm_count_map`, host with 0 VM from the same group resident on it is not inside.